napari_plugin_engine/validation.py

```diff
@@ -0,0 +1,125 @@
+import inspect
+import json
+import os
+from types import FunctionType
+from typing import Callable, Dict, List, Tuple, Union
+
+from napari.plugins import hook_specifications  # specification to be migrated
+from pkginfo import SDist, Wheel
+
+from napari_plugin_engine import HookImplementation, PluginManager
+
+functions: Dict[Tuple[str, str], Callable] = dict()
+
+
+def validate_artifacts(folder):
+    """
+    Validate the artifacts built for distribution:
+    raise AssertionError when the artifact is not build properly,
+
+    Parameters
+    ----------
+    folder : String
+        artifacts folder after build with setuptools
+    """
+    for pkgpath in os.listdir(folder):
+        pkgpath = os.path.join(folder, pkgpath)
+        if pkgpath.endswith('.tar.gz'):
+            dist = SDist(pkgpath)
+        elif pkgpath.endswith('.whl'):
+            dist = Wheel(pkgpath)
+        else:
+            print(f'Not a valid format {pkgpath}')
+            continue
+        assert 'Framework :: napari' in dist.classifiers
+        print(f'validated {pkgpath}')
+
+
+def validate_function(
```

Comment:
I think we should move the checks here because we want the developers to run these and found the issues as early as possible, seeing these error messages in napari by users would be too late and the users may not be able to figure out solutions anyway. We can continue polish on this and once we are comfortable, we remove the code in the napari core and call this instead

Reply:
I'd like us all to have a group conversation about what we're moving into napari plugin engine.  It does seem like the usage of NPE here is gradually changing from the original vision with each of these commits...  turning it from something that napari uses (but which the general community can definitely also use and benefit from), into something that would be weird for anyone but napari.  If you're surprised by that, and assumed this already napari specific, read the bullet points of features in the readme that discuss the generic benefits of this fork beyond pluggy: https://github.com/napari/napari-plugin-engine#napari-plugin-engine 

I'm not saying I'm 100% against that, but when I designed this repo in the beginning, I made a lot of effort to keep this generally applicable.  So things are starting to change with these commits.   It wasn't originally intended to just be a "light" library that people could import instead of depending on napari... (We assumed at first at least that they'd just depend on napari for those things).
So again, maybe that's what needs to happen, but before we lose all that isolation, can we discuss it more?

Reply:
totally welcome a group discussion, here's my 2cents:
1. I saw the original vision of making it more generally applicable, however, I don't see napari-plugin-engine would be reused by other groups, (probably dropping napari prefix would make others more comfortable referencing it), if the original goal was to extend pluggy so everyone could benefit, I would argue these changes should be proposed to pluggy instead of a fork of our own version if there is a universal use case.
2. Not sure if these are consensus, but what I felt is necessary is to have an API package, be it napari-plugin-engine or not, we would want to define the hook specifications in a separate place, so that plugin developers don't have to understand napari side changes to develop plugins, they would simply declare their plugin compatible with API version x,y,z and napari core would then declare support for different versions of API, instead of worrying about plugin compatibility.
3. One thing that I am hoping to achieve through this validation effort is to move the logic closer to developers, so that they would know whether a plugin would work or not at the time of development, instead of user trying to load the plugin and found that there was an error that they don't know how to deal with. In that principle, my argument is that all of these logics should live as part of the API package, once we decide where this API package should be, if not napari-plugin-engine

We can setup a focus group in our upcoming meeting to form a consensus, being completely honest I am also having doubt of what is the right thing to do. Sometimes I feel like I am overengineering and needs to be brought back to reality, it is not immediately clear how I can best serve the ongoing alfa cohort of developers and I could be overly trying to defensive design so they don't have to worry about anything

Reply:
> , I would argue these changes should be proposed to pluggy instead of a fork of our own version if there is a universal use case.

I agree.  we tried, and failed... couldn't get a response for quite a while (see issue 253 in that repo), pluggy devs were in disarray for a while and we needed some changes that weren't happening fast enough.  that was the reason for this fork.

> probably dropping napari prefix

that would be fine.  but i think my main point was: currently this "can" be used more generally... regardless of _how_ generally it is actually being used, we should be mindful of that, and discuss it before just changing that.

> 2. Not sure if these are consensus, but what I felt is necessary is to have an API package, be it napari-plugin-engine or not, we would want to define the hook specifications in a separate place, so that plugin developers don't have to understand napari side changes to develop plugins, they would simply declare their plugin compatible with API version x,y,z and napari core would then declare support for different versions of API, instead of worrying about plugin compatibility.

I recognize that this idea is very standard for web-type projects with REST/GraphQL APIs and schema.  I'm generally for it too (that was the original thing that drew us towards the hook spec idea in the first place).  but I'm not sure that there's a direct parallel (API version x,y,z) in the type of thing we're doing here.  The closest thing we have is the actual napari version... and for better or worse, it will likely take a _long_ time (if ever) to move to some sort of loosely coupled external API spec that napari obeys internally.  While I see that all over the place on the web, I really haven't seen it anywhere in python (happy to be shown otherwise! I think that pattern is intriguing...).  Anyway, while all these things sound great in theory, I just want to reiterate they they are somewhat grand design decisions that will take a lot of collaboration.  Personally, I'd like to see a bigger vision before we move code piecewise from one package to another.

> 3. One thing that I am hoping to achieve through this validation effort is to move the logic closer to developers, so that they would know whether a plugin would work or not at the time of development, instead of user trying to load the plugin and found that there was an error that they don't know how to deal with. In that principle, my argument is that all of these logics should live as part of the API package, once we decide where this API package should be, if not napari-plugin-engine

This is the bit that I think may well be impossible.  Even _if_ we had an API package that says what the function spec should be, I don't know why they wouldn't just test their package _with_ napari?  It's one thing to not have to _depend_ on napari to offer up plugin functionality (that's been an important concept all along).  but I'm not sure we need to extend that concept to testing a plugin too.  In order to "simulate" all of the workings of how napari uses plugins (without actually running napari), we'd really need to start duplicating a _ton_ of logic.  As I mentioned, it's super standard to say "when testing, install napari and do this"... so I'm still looking for a good reason why a package shouldn't want to install napari when testing the napari-specific part of their own package.


> We can setup a focus group in our upcoming meeting to form a consensus, being completely honest I am also having doubt of what is the right thing to do. Sometimes I feel like I am overengineering and needs to be brought back to reality, it is not immediately clear how I can best serve the ongoing alfa cohort of developers and I could be overly trying to defensive design so they don't have to worry about anything

that's very admirable!  I think we're all in agreement that we want the plugin developer experience to be smooth... but it also can't come at the expense of napari development itself.  It can be really tough to start spreading all of our critical bits into little side packages and keeping them all in sync.  So that's why I'd really like to see a pressing need.  Like, an example of something a plugin dev couldn't do by testing against napari proper (rather than some external testing package).  In that vein, I think we can probably offer up quite a bit of "convenience" by adding more fixtures to [`napari.utils._testsupport`](https://github.com/napari/napari/blob/master/napari/utils/_testsupport.py) (which is our "pytest plugin")... (as well as the github actions mentioned earlier)

Reply:
> we can probably offer up quite a bit of "convenience" by adding more fixtures to napari.utils._testsupport (which is our "pytest plugin")... (as well as the github actions mentioned earlier)

to provide a concrete examples here, imagine a `test_napari_widget` fixture.  It could actually try to add the widget to a viewer, select it from the dropdown menu, pass the viewer instance if requested, mock whatever isn't strictly necessary to the test at hand, etc...

then, a developer would use it as such (they _would_ of course need to add napari just to their test environment, but I'd say that's a good thing, as they can test against very specific napari versions):

```python
import pytest

def test_my_widget(test_napari_widget):
    from my_package import MyWidget
    test_napari_widget(MyWidget)
```